icekit/tests/tests.py

```diff
@@ -664,3 +674,174 @@ def test_join_words(self):
+        self.assertContains(response, self.listing.title)
+
+        # even if its parent listing is not published
+        # BOO - looks like fluent-pages won't resolve sub-urls of a page
+        # unless the page is published. Leaving commented code to indicate
+        # desired, but not necessary, functionality.
+        # self.article_2.publish()
+        # self.assertFalse(self.listing_2.has_been_published)
+        #
+        # response = self.app.get(self.article_2.get_absolute_url())
+        # self.assertNotContains(response, self.listing_2.title)
+        # self.article_2.unpublish()
+
+        self.listing.unpublish()
+        self.article.unpublish()
+
+    def test_publishing(self):
+        # check URLs 404 by default
+        self.listing_url = self.listing.get_absolute_url()
+        self.article_url = self.article.get_absolute_url()
+        response = self.app.get(self.listing_url, expect_errors=True)
+        self.assertEqual(response.status_code, 404)
+        response = self.app.get(self.article_url, expect_errors=True)
+        self.assertEqual(response.status_code, 404)
+
+        # Publish Article Listing
+        self.listing.publish()
+        self.listing_url = self.listing.get_published().get_absolute_url()
+
+        # Check it works, but doesn't include the Article
+        response = self.app.get(self.listing_url)
+        self.assertEqual(response.status_code, 200)
+
+        self.assertNotContains(response, self.article.title)
+
+        # Publish Article
+        self.article.publish()
+        self.article_url = self.article.get_published().get_absolute_url()
+        response = self.app.get(self.article_url)
+        self.assertEqual(response.status_code, 200)
+
+        # Check the article listing now includes the Article
+        response = self.app.get(
+            self.listing_url)
+        self.assertContains(response, self.article.title)
+
+        # Unpublish Article Listing
+        self.listing.unpublish()
+        response = self.app.get(self.listing_url, expect_errors=True)
+        self.assertEqual(response.status_code, 404)
+
+        # Unpublish Article to tidy up
+        self.article.unpublish()
+        response = self.app.get(self.article_url, expect_errors=True)
+        self.assertEqual(response.status_code, 404)
+
+    def test_listing(self):
+        # Check the listing doesn't include article_2, even if everything is
+        # published
+
+        self.article.publish()
+        self.article_2.publish()
+        self.listing.publish()
+
+        self.listing_url = self.listing.get_published().get_absolute_url()
+        self.article_url = self.article.get_published().get_absolute_url()
+
+        response = self.app.get(self.listing_url)
+        self.assertContains(response, self.article.title)
+        self.assertNotContains(response, self.article_2.title)
+
+        self.article.unpublish()
+        self.article_2.unpublish()
+        self.listing.unpublish()
```

Comment:
Test model implementations (and their migrations) and tests should all go in the app package being tested.